mne/_fiff/reference.py

```diff
@@ -87,6 +87,63 @@ def _check_before_reference(inst, ref_from, ref_to, ch_type):
     return ref_to
 
 
+def _check_before_dict_reference(inst, ref_dict):
```

Comment:
this function feels a bit long / inefficient. I think it could probably be simplified considerably. For example:

#### check that all dict keys are strings
```py
key_types = set(map(type, ref_dict))
assert len(key_types) == 1
assert next(iter(key_types)) is str
```

#### check that all values are str or list
```py
value_types = set(map(type, ref_dict.values()))
assert not value_types - {list, str}
```

#### check that all keys are valid channel names
```py
keys = set(ref_dict)
assert not keys - set(inst.ch_names)
```

#### check that all values are valid channel names
```py
values = {(x,) if isinstance(x, str) else tuple(x) for x in ref_dict.values()}
assert not set(sum(values, ())) - set(inst.ch_names)
```

(note that in MNE-Python we generally prefer comprehensions over `map()` and `filter()`, but the examples above should be straightforwardly translatable to comprehensions I think)

Reply:
... and also please don't use `assert ` for checking user-provided variables. In MNE-Python we only use them in code as "active comments" to assert things we should be *guaranteed already* to be true. This is because `assert` statements can be skipped by `python -O`. So also take these as shorthand for something more fleshed out, ideally with an informative error (e.g., if you provide a key that isn't a channel name, it would be nice to know what the errant key was).

It's okay to use `assert` statements in tests, though, because `pytest` overwrites the `assert` builtin.

Reply:
Thank you @drammock @larsoner! I have changed the assert to various kinds of errors.
I like the elegance of `map()` as it saves us from iterating keys and values at all, but a loop also allows us to more conveniently do channel type comparison, bad channel warning. Therefore I tried to re-organize the check a bit better while still retaining the loop structure. Hope it looks good enough now!

Reply:
I'm still a bit bothered by the nested functions and nested loops. Can you try something along these lines?  Not thoroughly tested but should be close:

```py
def _check_before_dict_reference(inst, ref_dict):
    """Prepare instance for dict-based referencing."""
    # Check to see that data is preloaded
    _check_preload(inst, "Applying a reference")

    # promote all values to list-like. this simplifies our logic and also helps catch
    # self-referencing cases like `{"Cz": ["Cz"]}`
    _refdict = {k: [v] if isinstance(v, str) else list(v) for k, v in ref_dict.items()}

    # check that keys are strings and values are (lists-of-)strings
    key_types = {type(k) for k in _refdict}
    value_types = {type(v) for val in _refdict.values() for v in val}
    for elem_name, elem in dict(key=key_types, value=value_types).items():
        if bad_elem := elem - {str}:
            raise TypeError(
                f"{elem_name.capitalize()}s in the ref_channels dict must be strings. "
                f"Your dict has {elem_name}s of type "
                f'{", ".join(map(lambda x: x.__name__, bad_elem))}.'
            )
    # check that keys are valid channels and values are (lists-of-)valid channels
    ch_set = set(inst.ch_names)
    keys = set(_refdict)
    values = {(x,) if isinstance(x, str) else tuple(x) for x in _refdict.values()}
    values = set(sum(values, ()))
    for elem_name, elem in dict(key=keys, value=values).items():
        if bad_elem := elem - ch_set:
            raise ValueError(
                f'ref_channels dict contains {elem_name}(s) ({", ".join(bad_elem)}) '
                "that are not names of channels in the instance."
            )
    # check for self-referencing
    self_ref = [[k] == v for k, v in _refdict.items()]
    if any(self_ref):
        which = np.array(list(_refdict))[np.nonzero(self_ref)]
        for ch in which:
            warn(f"Channel {ch} is self-referenced, which will nullify the channel")
    # check that channel types match
    ch_map = dict(zip(inst.ch_names, inst.get_channel_types()))
    pairs = [(k, v) for k in _refdict for v in _refdict[key]]  # unpack vals
    mismatch = [ch_map[k] != ch_map[v] for k, v in pairs]
    mismatch_pairs = np.array(pairs)[mismatch]
    if any(mismatch):
        for k, v in mismatch_pairs:
            warn(
                f"Channel {k} ({ch_map[k]}) is referenced to channel {v} which is a "
                f"different channel type ({ch_map[v]})"
            )
```

Reply:
Thanks for the feedback @drammock. I must say, I find the implementation you are suggesting a bit more difficult to read. It also doesn't seem to greatly shorten the code length. Are there other reasons to favor your implementation over ours? In our view, our implementation is a bit more friendly for anyone who would like to read and understand our code. Is there anything else you want us to modify?

Reply:
The main difference to my eye is code complexity (this is a technical term, related in part to the number of nested levels and parallel code paths AKA if/else branches). My suggested implementation has at most one level of for-loop and one level of if-clause (and no else-branches). To me it also seems (subjectively) to have a nicer flow: first check type of all entries, then check that the channels are actually in the object, then finally check for self referencing. IIRC your code does those checks for each channel as they are looped over. That means that a user might get an error for one entry, fix it, run their code again, and hit the same error on another entry. This can be frustrating --- better to be told *all* the non-conformant entries all at once so you can fix them all at once.

Caveat: all that is from memory... I'm on holiday and responding on a phone so it's hard for me to look closely again at the two implementations side-by-side. Thus I'll ask one of the other devs to weigh in and decide whether they agree that my suggested approach is substantively better, or just stylistically more to my taste. @hoechenberger @cbrnr @mscheltienne @larsoner could (at least) one of you have a look here and weigh in please?

Reply:
+1 for some variant of @drammock's suggestion -- it flattens out the code a bit, I find a bit easier to see what happens when and why, and should give more complete error messages/warnings